utilities/transactions/write_unprepared_txn.h

```diff
@@ -315,6 +319,16 @@ class WriteUnpreparedTxn : public WritePreparedTxn {
   // batch, it is possible that the delta iterator on the iterator will point to
   // invalid memory.
   std::vector<Iterator*> active_iterators_;
+
+  // Untracked keys that we have to rollback.
+  //
+  // TODO(lth): Currently we we do not record untracked keys per-savepoint.
+  // This means that when rolling back to savepoints, we have to check all
+  // keys in the current transaction for rollback. A more optimal solution
```

Comment:
I am confused how RollbackToSavepoint works. Would not this approach also rollback the keys before the save point?

Reply:
Yes it would, but that would only be inefficient, not incorrect because for RollbackToSavepoint, we do the rollback by reading the keys at the savepoint (not at the beginning of the transaction). This means that the rollback batch would just contain a reissue of the marker (ie. either a Put/Delete) and is not actually rolled back.

Reply:
Oh because of WriteUnpreparedTxnReadCallback? If yes, it would be helpful to add inline comments in two places: i) in RollbackToSavePointInternal to briefly the implication of using the callback, ii) here to mention the argument that you presented above.